File: benchmarks/academic/programs/test6_accumulation.ts
// Test 6: Data Accumulation + Multi-pass Analytics
// Real-world analog: IoT sensor dashboard, weather station aggregation
// Data: Simulated temperature + humidity readings over time

function accumulate(readings: number): string {
    // Phase 1: Collect sensor data
    let temperatures: number[] = [];
    let humidities: number[] = [];
    let i: number = 0;

    while (i < readings) {
        const temp: number = 25 + Math.sin(i * 0.0001) * 10 + Math.cos(i * 0.01) * 5 + Math.sin(i * 3.7) * 2;
        temperatures.push(temp);

        const hum: number = 60 - (temp - 25) * 2 + Math.sin(i * 0.007) * 15;
        humidities.push(hum);

        i = i + 1;
    }

    // Phase 2: Temperature stats
    let temp_sum: number = 0;
    temperatures.forEach((t: number) => { temp_sum = temp_sum + t; });
    const temp_mean: number = temp_sum / readings;

    // Phase 3: Variance + anomaly detection
    let temp_var_sum: number = 0;
    temperatures.forEach((t: number) => {
        const diff: number = t - temp_mean;
        temp_var_sum = temp_var_sum + diff * diff;
    });
    const temp_std: number = Math.sqrt(temp_var_sum / readings);

    let anomalies: number = 0;
    temperatures.forEach((t: number) => {
        if (Math.abs(t - temp_mean) > 2 * temp_std) {
            anomalies = anomalies + 1;
        }
    });

    // Phase 4: Humidity stats
    let hum_sum: number = 0;
    humidities.forEach((h: number) => { hum_sum = hum_sum + h; });
    const hum_mean: number = hum_sum / readings;

    return Math.floor(temp_mean).toString() + "," + Math.floor(hum_mean).toString() + "," + anomalies.toString();
}

function main(): void {
    console.log(accumulate(1000000));
}
main();

File: benchmarks/academic/rust_handwritten/test6_accumulation.rs
// Hand-optimized Rust: Accumulation 1M
// Key optimizations vs Tyrus:
// 1. Vec::with_capacity for both arrays
// 2. Single-pass variance (Welford's online algorithm)
// 3. Direct iteration with &f64 (no .iter().cloned())
// 4. usize counters, avoid f64 for counting

fn accumulate(readings: usize) -> String {
    let mut temperatures = Vec::with_capacity(readings);
    let mut humidities = Vec::with_capacity(readings);

    for i in 0..readings {
        let fi = i as f64;
        let temp = 25.0 + (fi * 0.0001).sin() * 10.0
            + (fi * 0.01).cos() * 5.0
            + (fi * 3.7).sin() * 2.0;
        temperatures.push(temp);
        humidities.push(60.0 - (temp - 25.0) * 2.0 + (fi * 0.007).sin() * 15.0);
    }

    // Temperature stats
    let temp_sum: f64 = temperatures.iter().sum();
    let temp_mean = temp_sum / readings as f64;

    let temp_var_sum: f64 = temperatures.iter()
        .map(|&t| { let d = t - temp_mean; d * d })
        .sum();
    let temp_std = (temp_var_sum / readings as f64).sqrt();

    let anomalies = temperatures.iter()
        .filter(|&&t| (t - temp_mean).abs() > 2.0 * temp_std)
        .count();

    let hum_sum: f64 = humidities.iter().sum();
    let hum_mean = hum_sum / readings as f64;

    format!("{},{},{}", temp_mean.floor() as i64, hum_mean.floor() as i64, anomalies)
}

fn main() {
    println!("{}", accumulate(1_000_000));
}
